### PR TITLE
feat(events): Add audit events payment confirm

### PR DESCRIPTION
**Pull request by Abhitator216**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
- Pass along request_state to payment_core
- Update `updatetrackers` trait to accept request state
- Update the paymentconfirm implementation of updatetracker to generate an event

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

- Generating domain events that can be used to track the high level overview of any process

### Test Case Description
- Implemented an event trigger that activates whenever the payment confirm API is utilized. This event triggers an audit event, the details of which are subsequently reflected in the Kafka UI.
- Create a payment & then confirm it 
- Fields that need to be tested are: `client_src`, `client_ver`and `frm_message`.
```
{"topic" = "hyperswitch-audit-events"}
```

image

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: config/development.toml
Comment (lsampras) [inline config/development.toml:1]: can you revert the changes for this file?

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 1]
```diff
@@ -1294,6 +1295,27 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
...
 
+        let client_src = header_payload
+            .client_source
+            .clone()
+            .or(payment_data.payment_attempt.client_source.clone());
+
+        let client_ver = header_payload
+            .client_version
+            .clone()
+            .or(payment_data.payment_attempt.client_version.clone());
```
Comment (lsampras) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1306]:
```suggestion
        let client_src = payment_data.payment_attempt.client_source.clone();

        let client_ver = payment_data.payment_attempt.client_version.clone();
```

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 2]
```diff
@@ -1294,6 +1295,27 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
...
+
+        let frm_message = payment_data.frm_message.clone();
+
+        _req_state
```
Comment (lsampras) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1310]:
```suggestion
        req_state
```

File: crates/router/src/core/payments/operations/payment_confirm.rs [hunk 3]
```diff
@@ -1294,6 +1295,27 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
...
+            .event_context
+            .event(AuditEvent::new(AuditEventType::PaymentConfirmed {
```
Comment (lsampras) [inline crates/router/src/core/payments/operations/payment_confirm.rs:1312]:
```suggestion
            .event(AuditEvent::new(AuditEventType::PaymentConfirm {
```

File: crates/router/src/events/audit_events.rs [hunk 4]
```diff
@@ -43,6 +52,7 @@ impl Event for AuditEvent {
         let event_type = match &self.event_type {
             AuditEventType::Error { .. } => "error",
             AuditEventType::PaymentCreated => "payment_created",
+            AuditEventType::PaymentConfirmed { .. } => "payment_confirmed",
```
Comment (lsampras) [inline crates/router/src/events/audit_events.rs:55]:
```suggestion
            AuditEventType::PaymentConfirm { .. } => "payment_confirm",
```

File: crates/router/src/events/audit_events.rs [hunk 5]
```diff
@@ -1,18 +1,27 @@
...
+        error_message: String,
+    },
     PaymentCreated,
     ConnectorDecided,
     ConnectorCalled,
     RefundCreated,
     RefundSuccess,
     RefundFail,
-    PaymentCancelled { cancellation_reason: Option<String> },
+    PaymentConfirmed {
```
Comment (lsampras) [inline crates/router/src/events/audit_events.rs:17]:
```suggestion
    PaymentConfirm {
```

**Review by lsampras — COMMENTED** [review]
minor nits

File: crates/router/src/core/payments/operations/payment_confirm.rs (hunk 2 above)
Comment (Abhitator216) [reply crates/router/src/core/payments/operations/payment_confirm.rs:1310]: _req_state is the only variable coming from above 
` _req_state: ReqState,`

**Review by lsampras — DISMISSED** [review]

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by ivor-juspay — APPROVED** [review]
